torch/jit.py

```diff
@@ -1,26 +1,375 @@
+            without deleting the actual `compile` invocation.
+
+    Example: Compile as function decorator.
+
+        >>> @jit.compile
+        >>> def f(x);
+        >>>     return x * 2
+        >>> x = Variable(torch.randn(1))
+        >>> out1 = f(x)  # interpreted run
+        >>> out1.sum().backward()  # required!
+        >>> out2 = f(x)  # compiled run
+        >>> out2.sum().backward()
+
+    Example: Compile as higher order function. (Notice that compile is a *curried*
+    function; you first apply it with the function/model to trace, and then
+    apply the result with the arguments.)
+
+        >>> compiled_model = jit.compile(nn.LSTMCell())
+        >>> out = compiled_model(input, hidden)
+
+    Example: Compile forwards only as function decorator
+
+        >>> @jit.compile(nderivs=0)
+        >>> def f(x);
+        >>>     return x * 2
+        >>> out1 = f(x)  # interpreted run
+        >>> out2 = f(x)  # compiled run
+
+    Example: Compile forwards only as higher order function
+
+        >>> compiled_model = jit.compile(nn.LSTMCell(), nderivs=0)
+        >>> out1 = compiled_model(input, hidden)  # interpreted run
+        >>> out2 = compiled_model(input, hidden)  # compiled run
+    """
+    # TODO: handle decorating a class (not an instance)
+    def _compile(inner):
+        if enabled:
+            return CompiledModule(inner, params=params, nderivs=nderivs, optimize=optimize)
+        else:
+            return inner
+    if callable(arg):
+        return _compile(arg)
+    else:
+        return _compile
+
+
+def trace(arg=None, nderivs=0, params=tuple()):
+    """
+    Instrument a function or module for tracing, wrapping it in a
+    :class:`TracedModule`, whose forward accepts the same arguments as the
+    original function/module, but returns a tuple consisting of the
+    *trace* of an execution, as well as the original return value.
+
+    Tracing is guaranteed not to change the semantics of the function/module
+    that is traced.
+
+    Arguments:
+        arg (optional, torch.nn.Module or function): the function or module
+            to be traced.  If `None`, `trace` returns a decorator which can be
+            applied to the function or module you want to trace.
+        nderivs (int, default 0): the number of derivatives to trace.
+            Traces of derivatives are recorded into the same trace returned
+            after executing the `forward` of the resulting module, but
+            are not present until you run `backward()` (an appropriate
+            number of times) on the resulting model.
+        params (tuple of torch.nn.Parameter): extra parameters for a traced
+            function, which do not occur as arguments to the function in
+            question.  You generally do not need this for tracing modules, as
+            the parameters of a module are automatically computed.
+
+    Example: Trace as function decorator.
+
+        >>> @jit.trace
```

Comment:
I am not sure I like this as an annotation. The common case will be recording a trace once.

Reply:
It's been expunged from the docs. It is still "supported" but we are not advertising it widely.